larpandoracontent/LArHelpers/LArHierarchyHelper.cc

```diff
@@ -0,0 +1,1171 @@
+/**
+ *  @file   larpandoracontent/LArHelpers/LArHierarchyHelper.cc
+ *
+ *  @brief  Implementation of the lar hierarchy helper class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/PdgTable.h"
+#include "Pandora/StatusCodes.h"
+
+#include "larpandoracontent/LArHelpers/LArHierarchyHelper.h"
+
+namespace lar_content
+{
+
+using namespace pandora;
+
+LArHierarchyHelper::MCHierarchy::MCHierarchy(const ReconstructabilityCriteria &recoCriteria) :
+    m_recoCriteria(recoCriteria),
+    m_pNeutrino{nullptr}
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+LArHierarchyHelper::MCHierarchy::~MCHierarchy()
+{
+    for (const Node *pNode : m_rootNodes)
+        delete pNode;
+    m_rootNodes.clear();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void LArHierarchyHelper::MCHierarchy::FillHierarchy(const MCParticleList &mcParticleList, const CaloHitList &caloHitList,
+    const bool foldToPrimaries, const bool foldToLeadingShowers)
+{
+    auto predicate = [](const MCParticle *pMC) { return std::abs(pMC->GetParticleId()) == NEUTRON; };
+    m_mcToHitsMap.clear();
+    for (const CaloHit *pCaloHit : caloHitList)
+    {
+        try
+        {
+            const MCParticle *pMC{MCParticleHelper::GetMainMCParticle(pCaloHit)};
+            m_mcToHitsMap[pMC].emplace_back(pCaloHit);
+        }
+        catch(const StatusCodeException&)
+        {
+        }
+    }
+
+    if (foldToPrimaries && !foldToLeadingShowers)
+    {
+        MCParticleSet primarySet;
+        m_pNeutrino = LArHierarchyHelper::GetMCPrimaries(mcParticleList, primarySet);
+        MCParticleList primaries(primarySet.begin(), primarySet.end());
+        if (m_recoCriteria.m_removeNeutrons)
+            primaries.erase(std::remove_if(primaries.begin(), primaries.end(), predicate), primaries.end());
```

Comment:
Are these first five lines the same in all of the four cases in the if / else statements? Could move them outside if that's the case.

Reply:
Yes, turns out they are - I think there's also one other set of code blocks relating to populating the joint hit lists that probably also warrants some refactoring.